test/Microsoft.SqlTools.ServiceLayer.IntegrationTests/SqlProjects/SqlProjectsServiceTests.cs

```diff
@@ -157,6 +159,95 @@ await service.HandleDeleteSqlObjectScriptRequest(new SqlProjectScriptParams()
             Assert.IsFalse(File.Exists(scriptFullPath), $"{scriptFullPath} expected to have been deleted from disk");
         }
 
+        [Test]
+        public async Task TestDatabaseReferenceAddDelete()
+        {
+            // Setup
+            SqlProjectsService service = new();
+            string projectUri = await service.CreateSqlProject();
+
+            SqlCmdVariable databaseVar = new SqlCmdVariable("$(OtherDb)", "OtherDbDefaultValue", "OtherDbValue");
+            SqlCmdVariable serverVar = new SqlCmdVariable("$(OtherServer)", "OtherServerDefaultValue", "OtherServerValue");
+
+            service.Projects[projectUri].SqlCmdVariables.Add(databaseVar);
+            service.Projects[projectUri].SqlCmdVariables.Add(serverVar);
```

Comment:
just to make sure I understand how this will work - ADS will be the one responsible for making the call to STS for adding any sqlcmd variables needed for a database reference? Will the database reference not get added if there's an error trying to add a sqlcmd var (like if the sqlcmd variable already exists) or will it still get added? Or do we need to add validation in ADS for that?

Reply:
> ADS will be the one responsible for making the call to STS for adding any sqlcmd variables needed for a database reference?

Correct.

> Will the database reference not get added if there's an error trying to add a sqlcmdvar?  Or do we need validation in ADS for that?

We'd need validation in ADS; the database reference will throw an `InvalidOperationException` when being added to a project if that project doesn't already contain a variable by that name.

---

Since the XML for database references only contain the name of the variable(s), but not the value or default value, there isn't enough information in _just_ the database reference to automatically add those variables, right?  Unless we wanted to do some sorta smart default sorta thing, but idk _what_ a smart default for `$(OtherServer)` and `$(OtherDatabase)` would even be...

Another possibility is that we could change the method signature to accept `SqlCmdVariable` objects for the database and server variables, and use those to add new vars if they aren't there.  That'd probably require us to add some new `Get()` functions to the collections (easy enough), but also to define the behavior if the passed variable matches on name but not value/default value (do we update? do we fail?).

Thoughts?

Reply:
The SSDT and ADS "Add database reference" dialogs get the sqlcmd variable name and value so that there's the information needed for adding a new sqlcmd variable. 

I also just noticed that the database variable isn't required, so if only database name is given, then `DatabaseVariableLiteralValue` is used instead of `DatabaseSqlCmdVariable`. Is this case also handled?
![image](https://user-images.githubusercontent.com/31145923/215851971-45d75fb8-4e1c-41f8-9e92-5fbc3b0ebcec.png)

Reply:
Err, right.  All of these cases should be covered, as long as the caller (whether that's STS directly or ADS via STS) creates the variables first.  I thought it made more sense to have ADS do two simple calls (one for variables, one for the reference) than one more complicated one that covers all the cases of both aspects.

Oh, shoot - I don't think it does.  I'll message you via teams to make sure I understand how that works.